src/python/pants/backend/native/config/environment.py

```diff
@@ -0,0 +1,84 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from abc import abstractproperty
+
+from pants.engine.rules import SingletonRule
+from pants.util.objects import datatype
+from pants.util.osutil import all_normalized_os_names, get_normalized_os_name
+
+
+class Platform(datatype(['normalized_os_name'])):
+
+  class UnsupportedPlatformError(Exception):
+    """Thrown if pants is running on an unrecognized platform."""
+
+  @classmethod
+  def create(cls):
+    return Platform(get_normalized_os_name())
+
+  _NORMALIZED_OS_NAMES = frozenset(all_normalized_os_names())
+
+  def resolve_platform_specific(self, platform_specific_funs):
```

Comment:
Should probably document this method, specifically what shape the argument should take.

Reply:
And the shape of the return value of course.

Reply:
Will do! #5933 may eventually turn into an alternative implementation of platform-specific logic, pending developments in creating subsystems and/or using variants in the v2 engine (neither of which I am currently familiar with), so making it clear why we're doing it this way now seems especially important.